src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// Provides a high-performance API for forward-only, read-only access to the UTF-8 encoded JSON text.
+    /// It processes the text sequentially with no caching and adheres strictly to the JSON RFC
+    /// by default (https://tools.ietf.org/html/rfc8259). When it encounters invalid JSON, it throws
+    /// a JsonReaderException with basic error information like line number and byte position on the line.
+    /// Since this type is a ref struct, it does not directly support async. However, it does provide
+    /// support for reentrancy to read incomplete data, and continue reading once more data is presented.
+    /// To be able to set max depth while reading OR allow skipping comments, create an instance of 
+    /// <see cref="JsonReaderState"/> and pass that in to the reader.
+    /// </summary>
+    public ref partial struct JsonUtf8Reader
+    {
+        private ReadOnlySpan<byte> _buffer;
+
+        private bool _isFinalBlock;
+
+        private ulong _stackFreeContainer;
+        private long _lineNumber;
+        private long _lineBytePosition;
+        private int _consumed;
+        private int _currentDepth;
+        private int _maxDepth;
+        private bool _inObject;
+        private bool _isNotPrimitive;
+        private JsonTokenType _tokenType;
+        private JsonReaderOptions _readerOptions;
+        private Stack<JsonTokenType> _stack;
+
+        private long _totalConsumed;
+        private bool _isLastSegment;
+        private readonly bool _isSingleSegment;
+        private SequencePosition _currentPosition;
+        //private SequencePosition _nextPosition; //TODO: Required for reading JSON from a ReadOnlySequeunce.
+        private ReadOnlySequence<byte> _sequence;
+
+        private bool IsLastSpan => _isFinalBlock && (_isSingleSegment || _isLastSegment);
+
+        /// <summary>
+        /// Lets the caller know which of the two 'Value' properties to read to get the 
+        /// token value. For input data within a ReadOnlySequence&lt;byte&gt; this will
+        /// always return false. For input data within a ReadOnlySequence&lt;byte&gt;, this
```

Comment:
For input data within a **Sequence** this will always return false.  For input data within a **Sequence**, this will only...

First ReadOnlySequence should be ReadOnlySpan.